gap/attributes/attract.gi

```diff
@@ -256,18 +256,32 @@ function(S)
 
   r := GreensRClassOfElementNC(S, rep);
 
-  if NrIdempotents(r) <> 1
-      or NrIdempotents(GreensDClassOfElementNC(S, rep)) <> 1 then
-    Info(InfoSemigroups, 2, "the number of idempotents in the D-class of the",
-         " first maximum rank");
-    Info(InfoSemigroups, 2, " generator is not 1");
+  if NrIdempotents(r) <> 1 or NrHClasses(r) <> 1 or
+      NrHClasses(GreensLClassOfElementNC(S, rep)) <> 1 then
+    Info(InfoSemigroups, 2, "the D-class of the first maximum rank generator ",
+                            "is not a group");
     return fail;
   fi;
 
   e := Idempotents(r)[1];
 
-  if ForAll(GeneratorsOfSemigroup(S), x -> x * e = x and e * x = x) then
-    return e;
+  if HasGeneratorsOfSemigroup(S) then
+    if ForAll(GeneratorsOfSemigroup(S), x -> x * e = x and e * x = x) then
+      return e;
```

Comment:
Why not combine these using `and` rather than having nested if's?

Reply:
Because if it does have generators of semigroup but doesn't satisfy the `ForAll`, then
it will proceed to perform the calculation using the `LambdaOrb` and `RhoOrb`.